util/tracer/tracer.go

```diff
@@ -0,0 +1,198 @@
+	// which they happened.
+	Content []TraceItem
+	tracer  *Tracer // origin tracer for clock, publishing...
+	depth   int32
+}
+
+// Event adds an Epoch with zero duration to the Trace.
+func (t *Trace) Event(name string) {
+	if t == nil {
+		return
+	}
+	t.epoch(name)()
+	t.Content[len(t.Content)-1].Duration = 0
+}
+
+// Epoch begins a phase in the life of the Trace, starting the measurement of
+// its duration. The returned function needs to be called when the measurement
+// is complete; failure to do so results in a panic() when Finalize() is
+// called. The suggested pattern of usage is, where possible,
+// `defer trace.Epoch("<name>")()`.
+func (t *Trace) Epoch(name string) func() {
+	if t == nil {
+		return func() {}
+	}
+	return t.epoch(name)
+}
+
+func (t *Trace) epoch(name string) func() {
+	if t.depth < 0 {
+		panic("use of finalized Trace:\n" + t.String())
+	}
+	t.depth++
+	pos := t.add(name)
+	called := false
+	return func() {
+		if called {
+			panic("epoch terminated twice")
+		}
+		called = true
+		t.Content[pos].Duration = t.tracer.now().Sub(t.Content[pos].Timestamp)
+		t.depth--
+	}
+}
+
+// Finalize submits the Trace to the underlying feed. If there is an open
+// Epoch, a panic occurs.
+func (t *Trace) Finalize() {
+	if t == nil {
+		return
+	}
+	if t.depth != 0 {
+		panic("attempt to finalize unbalanced trace:\n" + t.String())
+	}
+	t.depth = math.MinInt32
+	t.tracer.feed.Publish(t) // by reference
+}
+
+func (t *Trace) add(name string) int {
+	// Must be called with two callers to the client.
+	// (Client->Event|Epoch->epoch->add)
+	file, fun := t.tracer.callers.Lookup(3)
+	t.Content = append(t.Content, TraceItem{
+		depth:     t.depth,
+		Origin:    t.tracer.origin,
+		File:      file,
+		Func:      fun,
+		Timestamp: t.tracer.now(),
+		Name:      name,
+	})
+	return len(t.Content) - 1
+}
+
+// String implements fmt.Stringer. It prints a human-readable breakdown of the
+// Trace.
+func (t Trace) String() string {
+	const tab = "\t"
+	buf := bytes.NewBuffer(nil)
+	w := &tabwriter.Writer{}
+	w.Init(buf, 1, 1, 0, ' ', 0)
+	fmt.Fprintln(w, "Name", tab, "Origin", tab, "Ts", tab, "Dur", tab, "Desc", tab, "File")
+
+	const traceTimeFormat = "15:04:05.000000"
+	for _, c := range t.Content {
+		var namePrefix string
+		if c.depth > 1 {
+			namePrefix = strings.Repeat("·", int(c.depth-1))
+		}
+		fmt.Fprintln(w, t.Name, tab, c.Origin, tab,
+			c.Timestamp.Format(traceTimeFormat), tab, c.Duration, tab,
+			namePrefix+c.Name, tab, c.File)
+	}
+
+	_ = w.Flush()
+	return buf.String()
+}
+
+// A Tracer is used to follow requests across the system (or across systems).
+// Requests must implement the Traceable interface and can be traced by invoking
+// Create(), which returns a Trace object initialized to publish itself to a
+// util.Feed registered by the Tracer on completion.
+type Tracer struct {
+	origin  string // owner of this Tracer, i.e. Host ID
+	callers *CallResolver
```

Comment:
There's only one `Tracer` per Node, so it doesn't matter but I'll give it a try, always happy to get rid of a pointer. CallResolver contains a mutex, so that will probably have to turn into a pointer, which translates into more nil checks, though. We'll see.